File: Java/DSA/Sorting/QuickSort.java
package Sorting;

import java.util.Arrays;

public class QuickSort {

    public static void main(String[] args) {

        int[] arr = {10, 7, 8, 9, 1, 5};
        int n = arr.length;

        quickSort(arr, 0, n - 1);

        System.out.println("Sorted array:");
        System.out.println(Arrays.toString(arr));
    }

    static void quickSort(int[] arr, int l, int r) {
        if (l < r) {
            int p = partition(arr,l,r);

            quickSort(arr,l,p-1);
            quickSort(arr,p+1,r);
        }
    }

//    https://www.hackerearth.com/practice/algorithms/sorting/quick-sort/tutorial/
    private static int partition(int[] arr, int l, int r) {

        // here the partition will take such in place that all the elements before pivot will be less than it and after pivot will be greater than or equal to it.

        int piv = arr[r];
        int start = l-1;

        for(int j=l;j<r;j++){
            if(arr[j] < piv){
                start++;

                int temp = arr[j];
                arr[j] = arr[start];
                arr[start] = temp;
            }
        }

        int temp = arr[r];
        arr[r] = arr[start+1];
        arr[start+1] = temp;

        return start+1;
    }

}

File: Python/DSA/Sorting/quick_sort.py
def partition(arr, l, r):
    piv = arr[r]
    start = l - 1

    for i in range(l, r):
        if arr[i] < piv:
            start += 1
            arr[i], arr[start] = arr[start], arr[i]

    arr[start + 1], arr[r] = arr[r], arr[start + 1]
    return start + 1

def quick_sort(arr, l, r):
    if l < r:
        piv = partition(arr, l, r)
        quick_sort(arr, l, piv - 1)
        quick_sort(arr, piv + 1, r)
    
    return arr

# Example usage:
arr = [10, 7, 8, 9, 1, 5]
sorted_arr = quick_sort(arr, 0, len(arr) - 1)
print("Sorted array:", sorted_arr)
